src/lib/profiles/service-directory/ServiceDirectory.cpp

```diff
@@ -1133,79 +1165,20 @@ void WeaveServiceManager::onResponseReceived(uint32_t aProfileId, uint8_t aMsgTy
         VerifyOrExit(aMsgType == kMsgType_ServiceEndpointResponse, err = WEAVE_ERROR_INVALID_MESSAGE_TYPE);
         VerifyOrExit(mCacheState == kServiceMgrState_Waiting, err = WEAVE_ERROR_INCORRECT_STATE);
 
-        /*
-         * this block unpacks the service directory message.
-         */
+        err = unpackPacketBuffer(aMsg, true, &redir);
+        SuccessOrExit(err);
 
+#if WEAVE_CONFIG_PERSIST_SERVICE_DIRECTORY
+        WeaveLogProgress(ServiceDirectory, "label enabled");
+        if (!redir)
         {
-            MessageIterator i(aMsg);
-
-            uint16_t msgLen = aMsg->DataLength();
-            uint8_t dirCtrl;
-            uint8_t *writePtr;
-            uint8_t dirLen;
-            bool suffixesPresent;
-            uint8_t aLength;
-            bool timePresent = false;
-
-            err = i.readByte(&dirCtrl);
-            SuccessOrExit(err);
-
-            dirLen = dirCtrl & kMask_DirectoryLen;
-            redir = (dirCtrl & kMask_Redirect) != 0;
-            suffixesPresent = (dirCtrl & kMask_SuffixTablePresent) != 0;
-            timePresent = (dirCtrl & kMask_TimeFieldsPresent) != 0;
-
-            if (((msgLen > mCache.length) && !timePresent) || (msgLen > (mCache.length + (sizeof(uint64_t) + sizeof(uint32_t)))))
-            {
-                WeaveLogProgress(ServiceDirectory, "message length error: %d m.len:%d", msgLen, mCache.length);
-
-                err = WEAVE_ERROR_MESSAGE_TOO_LONG;
-            }
-            SuccessOrExit(err);
-
-            /*
-             * here we have directory information beyond the root directory but
-             * we're not done yet.
-             */
-
-            mDirectory.length = dirLen;
-            writePtr = mDirectory.base = mCache.base;
-
-            err = cacheDirectory(i, mDirectory.length, writePtr);
+            WeaveLogProgress(ServiceDirectory, "not redirecting, store service directory");
+            err = platform::StorePersistentServiceDir(aMsg->Start(),
```

Comment:
The behavior of this function is unclear.  If a copy of SD exists on disk when StorePersistentServiceDir() is called, but the version number passed to the function does not match the version stored on disk, is StorePersistentServiceDir() expected to clear the old version before writing the new version?

I would much prefer that this code called ClearPersistentServiceDir() before calling StorePersistentServiceDir().  In fact, I would prefer that it called ClearPersistentServiceDir() _before_ sending the service directory request.

Reply:
I called ClearPersistentServiceDir() in "connect", if state is "Initial". Please take a look.